nexus/db-queries/src/db/datastore/affinity.rs

```diff
@@ -368,28 +402,78 @@ impl DataStore {
-                    "Cannot paginate group members by name",
-                ));
-            }
+        let asc = match pagparams.direction {
+            dropshot::PaginationOrder::Ascending => true,
+            dropshot::PaginationOrder::Descending => false,
+        };
+
+        let mut query = QueryBuilder::new()
+            .sql(
+                "SELECT id,label FROM (
+                SELECT instance_id as id, 'instance' as label
+                FROM anti_affinity_group_instance_membership
+                WHERE group_id = ",
+            )
+            .param()
+            .bind::<diesel::sql_types::Uuid, _>(authz_anti_affinity_group.id())
+            .sql(
+                "
+                UNION
+                SELECT affinity_group_id as id, 'affinity_group' as label
+                FROM anti_affinity_group_affinity_membership
+                WHERE anti_affinity_group_id = ",
+            )
+            .param()
+            .bind::<diesel::sql_types::Uuid, _>(authz_anti_affinity_group.id())
+            .sql(") ");
+        if let Some(id) = pagparams.marker {
+            query = query
+                .sql("WHERE id ")
+                .sql(if asc { ">" } else { "<" })
+                .sql(" ")
+                .param()
+                .bind::<diesel::sql_types::Uuid, _>(*id);
+        };
+
+        query = query.sql(" ORDER BY id ");
+        if asc {
+            query = query.sql("ASC ");
+        } else {
+            query = query.sql("DESC ");
         }
-        .filter(dsl::group_id.eq(authz_anti_affinity_group.id()))
-        .select(AntiAffinityGroupInstanceMembership::as_select())
-        .load_async(&*self.pool_connection_authorized(opctx).await?)
-        .await
-        .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+
+        query =
+            query.sql(" LIMIT ").param().bind::<diesel::sql_types::BigInt, _>(
+                i64::from(pagparams.limit.get()),
+            );
+
+        query
+            .query::<(diesel::sql_types::Uuid, diesel::sql_types::Text)>()
+            .load_async::<(Uuid, String)>(
+                &*self.pool_connection_authorized(opctx).await?,
+            )
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?
+            .into_iter()
+            .map(|(id, label)| {
+                use external::AntiAffinityGroupMember as Member;
+                match label.as_str() {
+                    "affinity_group" => Ok(Member::AffinityGroup(
+                        AffinityGroupUuid::from_untyped_uuid(id),
+                    )),
+                    "instance" => Ok(Member::Instance(
+                        InstanceUuid::from_untyped_uuid(id),
+                    )),
+                    other => Err(external::Error::internal_error(&format!(
+                        "Unexpected label from database query: {other}"
+                    ))),
+                }
```

Comment:
this is kind of a silly idea, but: could we avoid the "other" case if we defined a SQL enum for these labels? not sure if it's actually worth the effort to do that, and having a type that doesn't actually appear in any tables is kinda weird, so take it or leave it.